Goals:
- Programatically (in python) download the appropriate gzip data from the link above one at a time.
- Unzip the data, and then delete the gzip file.
- Go through each csv and filter / clean out the appropriate data into a dataframe. This may require other libraries, such as reverse_geocoder.
- Delete the csvs when you are finished with them.
- Repeat steps 3-4 until you have a months worth of data, then transform that data to get the requested information above.
- Repeat steps 1-5 for each month and then for year between 1950 and 2000.
- At this point you should have a fully transformed dataset with yearly statistical data between 1950 and 2000.
- Export that data into a postgres database using sql alchemy.
- NOTE:  You will want to stop your existing container from running, then start a fresh database by making a new docker-compose file.  Ensure you have a .gitignore file so that the data on this postgres database isn't stored in git.



In [207]:
import numpy as np 
import pandas as pd
import requests
import tarfile
import re
import reverse_geocoder as rg
import glob 
import os

In [261]:
# Extract all files for 1950
for year in np.arange(1950, 2001):
    url = f"https://www.ncei.noaa.gov/data/global-summary-of-the-day/archive/{year}.tar.gz"
    response = requests.get(url, stream=True)
    file = tarfile.open(fileobj=response.raw, mode="r|gz")
    file.extractall(path=f'/Users/yorkmacbook030/ML_Cohort/Daily_Lessons/Weather_Agg_ETL/{year}')  # "."

In [262]:
# FUNCTION TO COMBINE CSV FILES INTO DATAFRAMES
def concat_csv_files(current_directory, years, file_pattern):
    
    df_results = [] #initialize blank list of dataframes
    
    for year in years:
        directory = os.path.join(current_directory, str(year))  #loop over folder for each year
        joined_files = os.path.join(directory, file_pattern)    #search for "*.csv" file patterns
        # print(joined_files)
        joined_list = glob.glob(joined_files)                   #add files to "joined_list"
        # print(joined_list)
        
        if joined_list:
            df_to_add = pd.concat(map(pd.read_csv, joined_list), ignore_index=True) #if file already in joined_list, add to it
            df_results.append(df_to_add)                                            #add df to df_results list

    df_final = pd.concat(df_results, ignore_index=True)
    
    return df_final


In [263]:
# Get working directory
current_directory = os.getcwd()
years = np.arange(1950, 2001)
# Concatenate files matching the pattern "7*.csv"
df_7 = concat_csv_files(current_directory, years, "7*.csv")

# # Concatenate files matching the pattern "69*.csv"
df_69 = concat_csv_files(current_directory, years, "69*.csv")

# # Concatenate files matching the pattern "99*.csv"
df_99 = concat_csv_files(current_directory, years, "99*.csv")

In [317]:
import shutil
# Function to delete all files in cwd
# List all files in the current working directory
files_in_cwd = os.listdir(current_directory)
# Delete all .csv files
for folder in files_in_cwd:
    if folder.isnumeric():
        file_path = os.path.join(current_directory, folder)
        shutil.rmtree(file_path)
    # if folder.startswith("0-9"):
    #     print(folder)
    #     file_path = os.path.join(directory_path, file_name)
    #     os.remove(file_path)
    # print(f"Deleted: {file_name}")

In [264]:
df_US = pd.concat([df_69,df_7,df_99])
df_US = df_US.dropna()
df_US.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22275774 entries, 0 to 4109351
Data columns (total 28 columns):
 #   Column            Dtype  
---  ------            -----  
 0   STATION           int64  
 1   DATE              object 
 2   LATITUDE          float64
 3   LONGITUDE         float64
 4   ELEVATION         float64
 5   NAME              object 
 6   TEMP              float64
 7   TEMP_ATTRIBUTES   int64  
 8   DEWP              float64
 9   DEWP_ATTRIBUTES   int64  
 10  SLP               float64
 11  SLP_ATTRIBUTES    int64  
 12  STP               float64
 13  STP_ATTRIBUTES    int64  
 14  VISIB             float64
 15  VISIB_ATTRIBUTES  int64  
 16  WDSP              float64
 17  WDSP_ATTRIBUTES   int64  
 18  MXSPD             float64
 19  GUST              float64
 20  MAX               float64
 21  MAX_ATTRIBUTES    object 
 22  MIN               float64
 23  MIN_ATTRIBUTES    object 
 24  PRCP              float64
 25  PRCP_ATTRIBUTES   object 
 26  SNDP              

In [265]:
df_US.head(2)

STATION        DATE  LATITUDE  LONGITUDE  ELEVATION  \
0  69011099999  1950-01-01    42.608    -82.835      177.0   
1  69011099999  1950-01-03    42.608    -82.835      177.0   

                          NAME  TEMP  TEMP_ATTRIBUTES  DEWP  DEWP_ATTRIBUTES  \
0  SELFRIDGE ANGB TRAIN, MI US  37.0                7  29.4                7   
1  SELFRIDGE ANGB TRAIN, MI US  15.2               17   1.7               17   

   ...  MXSPD   GUST   MAX  MAX_ATTRIBUTES   MIN  MIN_ATTRIBUTES  PRCP  \
0  ...    6.0  999.9  45.3               *  31.3               *   0.0   
1  ...   15.9  999.9  17.4               *   8.4               *   0.0   

   PRCP_ATTRIBUTES   SNDP  FRSHTT  
0                I  999.9       0  
1                I  999.9       0  

[2 rows x 28 columns]

#### Filter NAME field to exclude non-US country codes (Mexico, Canada, etc.), Alaska, and Hawaii.

In [266]:
#Check numbers of non_US, Hawaii, Alaska
non_US = df_US[df_US["NAME"].str.endswith("US")==False]
Hawaii = df_US["NAME"].str.endswith("HI US")
Alaska = df_US["NAME"].str.endswith("AK US")
print("Non-US names: ", len(non_US))
print("Hawaii: ", len(df_US[Hawaii] == True))
print("Alaska :", len(df_US[Alaska] == True))
print("Total: ", len(non_US) + len(df_US[Hawaii] == True) + len(df_US[Alaska] == True))


Non-US names:  6397441
Hawaii:  36380
Alaska : 1334072
Total:  7767893


In [267]:
df_US.shape

(22275774, 28)

In [268]:
# drop non-US, Hawaii, Alaska
US_yes = df_US["NAME"].str.endswith("US")
df_US = df_US[US_yes]
print("US only remaining: ", df_US.shape)
df_US = df_US[~df_US["NAME"].str.endswith("HI US")]
print("Without Hawaii remaining: ", df_US.shape)
df_US = df_US[~df_US["NAME"].str.endswith("AK US")]
print("Without Alaska remaining: ", df_US.shape)

US only remaining:  (15878333, 28)
Without Hawaii remaining:  (15841953, 28)
Without Alaska remaining:  (14507881, 28)


#### Filter using a latitude/longitude "net" to exclude areas within "US" that are outside the contiguous 48 states (ie. Puerto Rico)

In [269]:
#drop outside contiguous lat/long, ie. Puerto Rico
df_US = df_US[((df_US["LATITUDE"] > 24) & (df_US["LATITUDE"] < 50))
        & ((df_US["LONGITUDE"] < -66) & (df_US["LONGITUDE"] > -125))]
df_US.shape

(14314381, 28)

In [270]:
df_US.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14314381 entries, 0 to 4109351
Data columns (total 28 columns):
 #   Column            Dtype  
---  ------            -----  
 0   STATION           int64  
 1   DATE              object 
 2   LATITUDE          float64
 3   LONGITUDE         float64
 4   ELEVATION         float64
 5   NAME              object 
 6   TEMP              float64
 7   TEMP_ATTRIBUTES   int64  
 8   DEWP              float64
 9   DEWP_ATTRIBUTES   int64  
 10  SLP               float64
 11  SLP_ATTRIBUTES    int64  
 12  STP               float64
 13  STP_ATTRIBUTES    int64  
 14  VISIB             float64
 15  VISIB_ATTRIBUTES  int64  
 16  WDSP              float64
 17  WDSP_ATTRIBUTES   int64  
 18  MXSPD             float64
 19  GUST              float64
 20  MAX               float64
 21  MAX_ATTRIBUTES    object 
 22  MIN               float64
 23  MIN_ATTRIBUTES    object 
 24  PRCP              float64
 25  PRCP_ATTRIBUTES   object 
 26  SNDP              

### df_US is now cleaned and filtered to only US locations from the lower 48 states with no missing values

In [271]:
#ensure "DATE" columns follow datetime format so we can group by month
df_US["DATE"] = pd.to_datetime(df_US["DATE"])

In [272]:
# Create DATE & MONTH columns for final aggregation
df_US["MONTH"] = df_US["DATE"].dt.month
df_US["YEAR"] = df_US["DATE"].dt.year

#### Now let's explore missing temperature and precipitation data.

In [273]:
print("Missing max values: ", round((len(df_US[df_US["MAX"] == 9999.9])/len(df_US)*100),2), "%")   #Missing TEMP,MAX,MIN values are entered as 99.99
print("Missing min values: ", round((len(df_US[df_US["MIN"] == 9999.9])/len(df_US)*100),2), "%")
print("Missing temp values: ", len(df_US[df_US["TEMP"] == 9999.9])/len(df_US)*100, "%")
print("Missing prcp values: ", round((len(df_US[df_US["PRCP"] == 99.99])/len(df_US)*100),2), "%")  #Missing PRCP values are entered as 99.99
print("PRCP '0' Values: ", round((len(df_US[df_US["PRCP"] == 0])/len(df_US)*100),2), "%")

Missing max values:  0.14 %
Missing min values:  0.12 %
Missing temp values:  0.0 %
Missing prcp values:  13.55 %
PRCP '0' Values:  72.83 %


##### 11.46% of precipitation values were not entered, and 71.66% of precipitation values are equal to 0.  According to the documentation, it is likely that some or many of the 99.99 values could be assumed to mean no precipitation fell that day, although non-zero precipitation could be entered as 99.99. Even so, to avoid introduction of too much bias, I will break the data set into 4 separate sets, drop the missing values and perform the aggregated statistical analysis separately.  I will recombine the four sets together at the end.  

In [274]:
#Create 4 separate df's with only year, month, and weather statistic needed
TEMP = df_US[["YEAR", "MONTH", "TEMP"]]
MAX = df_US[["YEAR", "MONTH", "MAX"]]
MIN = df_US[["YEAR", "MONTH", "MIN"]]
PRCP = df_US[["YEAR", "MONTH", "PRCP"]]

In [275]:
len(TEMP[TEMP["TEMP"]==9999.9])

0

##### TEMP has 0 missing values

In [276]:
# TEMP has no missing values, so just aggregate by year/month:
min = TEMP.groupby(["YEAR", "MONTH"]).min()
print(len(min))
max = TEMP.groupby(["YEAR", "MONTH"]).max()
print(len(min))
avg = TEMP.groupby(["YEAR", "MONTH"]).mean()
print(len(min))
stdev = TEMP.groupby(["YEAR", "MONTH"]).std()
print(len(min))

612
612
612
612


In [277]:
TEMP_final = pd.concat([min,max,avg,stdev],axis=1)
TEMP_final.columns=["Abs_min","Abs_max","Abs_avg","Abs_stdev"]
TEMP_final.head(2)

Abs_min  Abs_max    Abs_avg  Abs_stdev
YEAR MONTH                                        
1950 1        -36.1     76.8  37.570788  20.396160
     2        -21.4     77.4  39.525718  16.251752

##### Now look at MAX data.

In [278]:
len(MAX[MAX["MAX"] == 9999.9])

19465

##### Max has 19465 missing values, which is less than 1% of the data, so I'll drop them.

In [279]:
print(MAX.shape)
drop_list = MAX[MAX["MAX"]==9999.9].index
MAX = MAX.drop(drop_list)
print(MAX.shape)

(14314381, 3)
(14291339, 3)


In [280]:
min = MAX.groupby(["YEAR", "MONTH"]).min()
print(len(min))
max = MAX.groupby(["YEAR", "MONTH"]).max()
print(len(min))
avg = MAX.groupby(["YEAR", "MONTH"]).mean()
print(len(min))
stdev = MAX.groupby(["YEAR", "MONTH"]).std()
print(len(min))

612
612
612
612


In [281]:
MAX_final = pd.concat([min, max, avg, stdev], axis=1)
MAX_final.columns = ["MAX_min", "MAX_max", "MAX_avg", "MAX_stdev"]
MAX_final.head(2)

MAX_min  MAX_max    MAX_avg  MAX_stdev
YEAR MONTH                                        
1950 1        -24.0     93.4  48.139187  20.631453
     2        -11.9     92.3  50.482757  17.280712

##### Now look at MIN data

In [282]:
len(MIN[MIN["MIN"] == 9999.9])

17546

##### MIN has 17546 missing values, which is less than 1% of the data, so I'll drop them.

In [283]:
print(MIN.shape)
drop_list = MIN[MIN["MIN"]==9999.9].index
MIN = MIN.drop(drop_list)
print(MIN.shape)

(14314381, 3)
(14294289, 3)


In [284]:
min = MIN.groupby(["YEAR", "MONTH"]).min()
print(len(min))
max = MIN.groupby(["YEAR", "MONTH"]).max()
print(len(min))
avg = MIN.groupby(["YEAR", "MONTH"]).mean()
print(len(min))
stdev = MIN.groupby(["YEAR", "MONTH"]).std()
print(len(min))

612
612
612
612


In [285]:
MIN_final = pd.concat([min, max, avg, stdev], axis=1)
MIN_final.columns = ["MIN_min", "MIN_max", "MIN_avg", "MIN_stdev"]
MIN_final.head(2)

MIN_min  MIN_max   MIN_avg  MIN_stdev
YEAR MONTH                                       
1950 1        -50.1     73.9  29.31605  21.200445
     2        -38.9     73.0  30.87958  16.761127

##### Now look at PRCP data

In [286]:
len(PRCP[PRCP["PRCP"] == 99.99])

1939055

##### PRCP has 1939055 missing values, which is about than 14% of the data.  Although it is fair to assume that many of these correspond to actual 0 precipitation days, I will drop them because we still have a large dataset.

In [287]:
print(PRCP.shape)
drop_list = PRCP[PRCP["PRCP"]==99.99].index
PRCP = PRCP.drop(drop_list)
print(PRCP.shape)

(14314381, 3)
(11563283, 3)


In [293]:
min = PRCP.groupby(["YEAR", "MONTH"]).min()
print(len(min))
max = PRCP.groupby(["YEAR", "MONTH"]).max()
print(len(max))
avg = PRCP.groupby(["YEAR", "MONTH"]).mean()
print(len(avg))
stdev = PRCP.groupby(["YEAR", "MONTH"]).std()
print(len(stdev))

612
612
612
612


In [294]:
PRCP_final = pd.concat([min, max, avg, stdev], axis=1)
PRCP_final.columns = ["PRCP_min", "PRCP_max", "PRCP_avg", "PRCP_stdev"]
PRCP_final.head(20)

PRCP_min  PRCP_max  PRCP_avg  PRCP_stdev
YEAR MONTH                                          
1950 1           0.0      2.68  0.010894    0.086276
     2           0.0      1.65  0.007122    0.068520
     3           0.0      2.09  0.010149    0.083979
     4           0.0      2.44  0.007201    0.066167
     5           0.0      4.53  0.007128    0.083491
     6           0.0      2.60  0.005822    0.070303
     7           0.0      3.27  0.010982    0.113733
     8           0.0      3.39  0.006641    0.086005
     9           0.0      9.76  0.009233    0.142612
     10          0.0      2.87  0.004570    0.065539
     11          0.0      2.91  0.006565    0.075579
     12          0.0      2.60  0.007059    0.075504
1951 1           0.0      1.14  0.006138    0.055552
     2           0.0      2.99  0.008873    0.085004
     3           0.0      1.89  0.008328    0.069832
     4           0.0      2.01  0.007438    0.065895
     5           0.0      2.01  0.005362    0.060401
     6           0.0      2.32  0.009444    0.089149
     7           0.0      2.36  0.006879    0.078516
     8           0.0      2.87  0.006552    0.078730

In [301]:
df_out = pd.concat(
    [
        TEMP_final,
        MAX_final,
        MIN_final,
        PRCP_final
    ],
    axis=1,
)
display(df_out.head(2))
print("Dimensions of df_out: ",df_out.shape)
df_out.head(100)

Abs_min  Abs_max    Abs_avg  Abs_stdev  MAX_min  MAX_max  \
YEAR MONTH                                                             
1950 1        -36.1     76.8  37.570788  20.396160    -24.0     93.4   
     2        -21.4     77.4  39.525718  16.251752    -11.9     92.3   

              MAX_avg  MAX_stdev  MIN_min  MIN_max   MIN_avg  MIN_stdev  \
YEAR MONTH                                                                
1950 1      48.139187  20.631453    -50.1     73.9  29.31605  21.200445   
     2      50.482757  17.280712    -38.9     73.0  30.87958  16.761127   

            PRCP_min  PRCP_max  PRCP_avg  PRCP_stdev  
YEAR MONTH                                            
1950 1           0.0      2.68  0.010894    0.086276  
     2           0.0      1.65  0.007122    0.068520

Dimensions of df_out:  (612, 16)


Abs_min  Abs_max    Abs_avg  Abs_stdev  MAX_min  MAX_max  \
YEAR MONTH                                                             
1950 1        -36.1     76.8  37.570788  20.396160    -24.0     93.4   
     2        -21.4     77.4  39.525718  16.251752    -11.9     92.3   
     3        -13.7     84.9  42.843263  15.198640     -2.0    105.1   
     4         10.2     86.4  51.881509  12.907372     18.0    105.1   
     5         19.1     94.4  62.568186  11.340779     23.0    109.9   
...             ...      ...        ...        ...      ...      ...   
1957 12       -14.4     77.1  42.728488  12.617666     -5.1     84.9   
1958 1         -8.1     74.2  36.997851  12.694950      3.0     82.4   
     2        -17.4     77.1  36.749190  16.571451     -4.0     88.0   
     3          0.2     78.2  43.540616  11.501368     10.0     90.3   
     4         16.9     89.0  55.367415  11.404861     28.0    109.2   

              MAX_avg  MAX_stdev  MIN_min  MIN_max    MIN_avg  MIN_stdev  \
YEAR MONTH                                                                 
1950 1      48.139187  20.631453    -50.1     73.9  29.316050  21.200445   
     2      50.482757  17.280712    -38.9     73.0  30.879580  16.761127   
     3      54.267682  16.706826    -29.0     75.9  33.668238  15.078124   
     4      63.882381  14.618549    -13.0     77.0  41.848190  12.610145   
     5      74.265632  12.246296     12.0     80.1  52.437396  11.886272   
...               ...        ...      ...      ...        ...        ...   
1957 12     52.919534  13.335772    -25.1     72.0  34.581670  13.289317   
1958 1      46.525708  13.533526    -29.9     71.1  29.487919  13.193313   
     2      46.316340  17.598401    -31.0     73.0  29.164206  16.787736   
     3      52.322020  12.696250    -11.9     75.0  36.469514  11.591063   
     4      66.333030  12.516209      9.0     81.0  46.023778  11.682461   

            PRCP_min  PRCP_max  PRCP_avg  PRCP_stdev  
YEAR MONTH                                            
1950 1           0.0      2.68  0.010894    0.086276  
     2           0.0      1.65  0.007122    0.068520  
     3           0.0      2.09  0.010149    0.083979  
     4           0.0      2.44  0.007201    0.066167  
     5           0.0      4.53  0.007128    0.083491  
...              ...       ...       ...         ...  
1957 12          0.0      2.52  0.009915    0.083656  
1958 1           0.0      2.64  0.012685    0.102926  
     2           0.0      2.40  0.011969    0.093068  
     3           0.0      2.28  0.015068    0.104445  
     4           0.0      4.09  0.014423    0.110910  

[100 rows x 16 columns]

In [320]:
import psycopg2
import sqlalchemy

# Create the engine to connect to the PostgreSQL database
engine = sqlalchemy.create_engine(
    "postgresql://postgres:password@localhost:5432/postgres"
)

# Write data into the table in PostgreSQL database
with engine.begin() as connection:
    df_out.to_sql("Luke_Sands_Weather_ETL", con=connection)

In [326]:
conn = psycopg2.connect(
    database="postgres",
    host="localhost",
    user="postgres",
    password="password",
    port=5432,
)

In [327]:
cursor = conn.cursor()

In [328]:
cursor.execute("SELECT * FROM Luke_Sands_Weather_ETL")
rows = cursor.fetchall()

UndefinedTable: relation "luke_sands_weather_etl" does not exist
LINE 1: SELECT * FROM Luke_Sands_Weather_ETL
                      ^
